vcs/src/main/java/org/openjdk/skara/vcs/git/GitRepository.java

```diff
@@ -1659,4 +1659,65 @@ public int commitCount() throws IOException {
     public Hash initialHash() {
         return EMPTY_TREE;
     }
+
+    @Override
+    public Optional<Hash> wholeHash(String rev) {
+        try (var p = capture("git", "rev-parse", rev)) {
+            var res = p.await();
+            if (res.status() == 0 && res.stdout().size() == 1) {
+                return Optional.of(new Hash(res.stdout().get(0)));
+            }
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public Optional<List<String>> stagedFile(Path path) {
+        try (var p = capture("git", "cat-file", "-p", ":" + path.toString())) {
+            var res = p.await();
+            if (res.status() == 0) {
+                return Optional.of(res.stdout());
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Commit staged() throws IOException {
+        var author = new Author("jcheck", "jcheck@none.none");
```

Comment:
For the default user and email, there is already a `::userName` method. You can create one for email as well, and then use these values as fallbacks (`Optional.orElse`).

Reply:
> There are a lot of minor things, but all in all this turned out better than I had imagined. Looking forward to using it.
> 
> Have you tried all the options against a repository with various degrees of local changes in it?

Yes,  I tried all the options and all the combination of options.